### PR TITLE
Use redux thunk for handling async code

### PR DESCRIPTION
**Pull request by bradr5**
# Redux Thunks

## Background

Sagas are our current redux middleware used for writing async functions. Although they can be very powerful, they're overkill in most instances and bring about a higher learning curve than other alternatives. Sagas are good for complex async workflows that involve things such as debouncing, cancelling requests, and background threaded behavior – benefits that are generally not needed in our use case.

## Description

In an attempt to remove sagas from our code this approach utilizes redux thunks as a middleware for writing async logic as basic functions

Pros

* Components don't have to be aware of how actions are implemented (whether synchronous or asynchronous)
* Easy to maintain especially when multiple components require the same async logic for updates redux state
* Officially recommended by redux

Cons

* Somewhat ambiguous not knowing what the action is doing when looking from the react side
* Might make mocking for unit tests a little more difficult

### CODE REVIEW

File: ui/apps/platform/src/Containers/SystemConfig/Form/SystemConfigForm.tsx [hunk 1]
```diff
@@ -96,13 +96,12 @@ const SystemConfigForm = ({
                         });
                         setSystemConfig(data);
                         setErrorMessage(null);
+                        setSubmitting(false);
```
Comment (bradr5) [inline ui/apps/platform/src/Containers/SystemConfig/Form/SystemConfigForm.tsx:99]: Changes in this file are unrelated to the redux change. A state change (`setSubmitting(false)`) was happening after the component was destroyed due to `setIsNotEditing()`

File: ui/apps/platform/src/index.tsx [hunk 2]
```diff
@@ -44,6 +47,14 @@ const history = createHistory();
 const store = configureStore(undefined, history) as Store;
 const apolloClient = configureApollo();
 
+const dispatch = (action) =>
```
[the remark this answers is not in the document]
Comment (bradr5) [reply ui/apps/platform/src/index.tsx:50]: Yeah unfortunately this is the side effect of not having the store properly typed in configureStore as well as not having our redux state and reducers currently typed out. We have a few options to get around this ugly setup and I'm trying them out in a separate stash right now

File: ui/apps/platform/src/reducers/featureFlags.js [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
 
 export const types = {
     FETCH_FEATURE_FLAGS: createFetchingActionTypes('featureflags/FETCH_FEATURE_FLAGS'),
 };
 
-export const actions = {
-    fetchFeatureFlags: createFetchingActions(types.FETCH_FEATURE_FLAGS),
+// Actions
+
+export const fetchFeatureFlags = () => {
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/featureFlags.js:14]: Yes, reducer actions are parallel to containers which have 3 `useState` hooks:
* request: similar to `setIsFetchingWhatever(true)`
* success: similar to `setWhatever(data)` and `setIsFetchingWhatever(false)`
* failure: similar to `setWhateverError(e)` and `setIsFetchingWhatever(false)`

Reducers will provide 3 properties of state: isFetching/isLoading, data, error (not dictating these names)

File: ui/apps/platform/src/reducers/systemConfig.js [hunk 4]
```diff
@@ -11,8 +12,20 @@ export const types = {
 
 // Actions
 
-export const actions = {
-    fetchPublicConfig: createFetchingActions(types.FETCH_PUBLIC_CONFIG),
+export const fetchPublicConfig = () => {
```
Comment (pedrottimark) [inline ui/apps/platform/src/reducers/systemConfig.js:15]: A question, not necessarily a change request, is there a variation on `fetchPublicConfig` name?
* which removes the need to alias `fetchPublicConfig` service function
* which reduces possible confusion between the action and service functions for Find in Files (not a new problem: yesterday, because of same name for the pair of functions, I had to analyze each occurrence to confirm that the code no longer calls an action function, nor calls the service function directly, therefore the indirect call via sagas is obsolete).

File: ui/apps/platform/src/reducers/systemConfig.js [hunk 5]
```diff
@@ -11,8 +12,20 @@ export const types = {
...
+        dispatch({ type: types.FETCH_PUBLIC_CONFIG.REQUEST });
+
+        try {
+            const result = await fetchPublicConfigService();
+            dispatch({
+                type: types.FETCH_PUBLIC_CONFIG.SUCCESS,
+                response: result.response,
+            });
+        } catch (e) {
+            dispatch({ type: types.FETCH_PUBLIC_CONFIG.FAILURE, payload: e });
```
Comment (pedrottimark) [inline ui/apps/platform/src/reducers/systemConfig.js:26]: What do you think about the following?

```js
        } catch (error) {
            dispatch({ type: types.FETCH_PUBLIC_CONFIG.FAILURE, error });
```

File: ui/apps/platform/src/reducers/systemConfig.js [hunk 6]
```diff
@@ -11,8 +12,20 @@ export const types = {
...
-    fetchPublicConfig: createFetchingActions(types.FETCH_PUBLIC_CONFIG),
+export const fetchPublicConfig = () => {
+    return async (dispatch) => {
+        dispatch({ type: types.FETCH_PUBLIC_CONFIG.REQUEST });
+
+        try {
+            const result = await fetchPublicConfigService();
+            dispatch({
+                type: types.FETCH_PUBLIC_CONFIG.SUCCESS,
+                response: result.response,
```
Comment (pedrottimark) [inline ui/apps/platform/src/reducers/systemConfig.js:23]: Action object here and in SystemConfigForm.tsx file might become simpler with possible future change:
* from
    ```js
    export function fetchPublicConfig(): Promise<{ response: PublicConfig }> {
        return axios.get<PublicConfig>(`${baseUrl}/public`).then(({ data }) => ({
            response: data,
        }));
    }
    ```
* to
    ```js
    export function fetchPublicConfig(): Promise<PublicConfig> {
        return axios.get<PublicConfig>(`${baseUrl}/public`).then((response) => response.data);
    }
    ```

File: ui/apps/platform/src/reducers/featureFlags.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/featureFlags.js:14]: Same thought question here as below. Pro and con of different names for action and service functions?

[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/featureFlags.js:14]: Also similar possible simplifications as below if future change in service function:
* from
    ```js
    export function fetchFeatureFlags(): Promise<{ response: { featureFlags: FeatureFlag[] } }> {
        return axios.get<{ featureFlags: FeatureFlag[] }>(url).then((response) => ({
            response: response.data,
        }));
    }
    ```
* to
    ```js
    export function fetchFeatureFlags(): Promise<FeatureFlag[]> {
        return axios.get<{ featureFlags: FeatureFlag[] }>(url).then((response) => response.data.featureFlags;
    }
    ```

File: ui/apps/platform/src/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/index.tsx:50]: A future follow to rewrite the forward-looking reducers in TypeScript might help here?

File: ui/apps/platform/src/reducers/systemConfig.js (hunk 4 above)
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/systemConfig.js:15]: Thinking a bit more, without revealing too much about the implementation, this action function, and corresponding action functions for other forward-looking reducers, are only for `dispatch` call at top level, correct? So a naming convention might help communicate that they:
* are similar to each other
* are different from the other pure action functions in the reducer

**Review by vjwilson — COMMENTED** [review]
I had a couple of questions about parameter names.

File: ui/apps/platform/src/reducers/featureFlags.js [hunk 7]
```diff
@@ -1,18 +1,37 @@
...
+        dispatch({ type: types.FETCH_FEATURE_FLAGS.REQUEST });
+
+        try {
+            const result = await fetchFeatureFlags();
+            dispatch({
+                type: types.FETCH_FEATURE_FLAGS.SUCCESS,
+                response: result.response,
+            });
+        } catch (error) {
+            dispatch({ type: types.FETCH_FEATURE_FLAGS.FAILURE, payload: error });
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/featureFlags.js:25]: Good eyes. Especially because of comment below, here is a clear and simple naming convention:

```js
dispatch({ type: types.FETCH_FEATURE_FLAGS.FAILURE, error });
```

[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/reducers/featureFlags.js:25]: We kept `response` from original code to limit initial scope to minimum changes.

A possible future improvement is to rename `response` property as either of the following:
* generic `data`  to undo confusing convention of service functions for sagas `(response) => ({ response: response.data })`
* specific `featureFlags` to eliminate extra level of property access, like in reducer below:
    `action.response.featureFlags` which could be `action.featureFlags` because we do not expect additional properties in the response of such a mature API endpoint

File: ui/apps/platform/src/reducers/systemConfig.js [hunk 8]
```diff
@@ -24,16 +37,51 @@ const publicConfig = (state = {}, action) => {
...
 };
 
+const error = (state = null, action) => {
+    switch (action.type) {
+        case types.FETCH_PUBLIC_CONFIG.REQUEST:
+        case types.FETCH_PUBLIC_CONFIG.SUCCESS:
+            return null;
+
+        case types.FETCH_PUBLIC_CONFIG.FAILURE:
+            return action.error;
```
[the remark this answers is not in the document]
Comment (bradr5) [reply ui/apps/platform/src/reducers/systemConfig.js:47]: Good catch, this and feature flags have been updated

File: ui/apps/platform/src/reducers/systemConfig.js (hunk 4 above)
Comment (bradr5) [reply ui/apps/platform/src/reducers/systemConfig.js]: I mostly mimicked the pattern from other reducers but I agree. I updated multiple function names to prevent needing to alias

**Review by vjwilson — APPROVED** [review]

**Comment by bradr5** [comment]
Due to this work, authentication tests are being skipped (can be found in [PR 3997](https://github.com/stackrox/stackrox/pull/3997)). Previously, errors in the feature flags reducer file caused integration tests to deceptively pass after the login screen due to this code in `MainPage.tsx` where the app would get stuck on the loading screen:

```javascript
// Render Body and NavigationSideBar only when feature flags and permissions are available.
if (isLoadingFeatureFlags || isLoadingPermissions) {
    return <LoadingSection message="Loading..." />;
}
```

After correcting the issue in the feature flags reducer, authentication tests began failing because required redux properties past the loading screen were not getting setup properly.
